src/engine/structures/lock.rs

```diff
@@ -993,35 +1189,205 @@ where
+    }
+}
+
+impl<U, T> AllLockWriteGuard<U, T>
+where
+    U: AsUuid,
+    T: ?Sized,
+{
+    pub fn get_by_uuid(&self, u: U) -> Option<(Name, &T)> {
+        self.1
+            .get_by_uuid(u)
+            .map(|(n, p)| (n, unsafe { p.as_ref().expect("Not null") }))
+    }
+
+    pub fn get_by_name(&self, name: &Name) -> Option<(U, &T)> {
+        self.1
+            .get_by_name(name)
+            .map(|(u, p)| (u, unsafe { p.as_ref().expect("Not null") }))
+    }
+
+    pub fn get_mut_by_uuid(&mut self, u: U) -> Option<(Name, &mut T)> {
+        self.1
+            .get_by_uuid(u)
+            .map(|(n, p)| (n, unsafe { p.as_mut().expect("Not null") }))
+    }
+
+    pub fn get_mut_by_name(&mut self, name: &Name) -> Option<(U, &mut T)> {
+        self.1
+            .get_by_name(name)
+            .map(|(u, p)| (u, unsafe { p.as_mut().expect("Not null") }))
+    }
+
+    pub fn iter(&self) -> AllLockWriteGuardIter<'_, U, T> {
+        AllLockWriteGuardIter(self.1.iter())
+    }
+
+    pub fn iter_mut(&mut self) -> AllLockWriteGuardIterMut<'_, U, T> {
+        AllLockWriteGuardIterMut(self.1.iter_mut())
+    }
+
+    #[cfg(test)]
+    pub fn len(&self) -> usize {
+        self.1.len()
+    }
+}
+
+impl<U, T> Drop for AllLockWriteGuard<U, T>
+where
+    U: AsUuid,
+    T: ?Sized,
+{
+    fn drop(&mut self) {
+        trace!("Dropping all write lock");
+        if self.2 {
+            let mut lock_record = self.0.lock().expect("Mutex only locked internally");
+            lock_record.remove_write_all_lock();
+            lock_record.wake();
+        }
+        trace!("All write lock dropped");
+    }
+}
+
+/// Future returned by AllOrSomeLock::write_all().
+struct AllModify<U: AsUuid, T>(AllOrSomeLock<U, T>, AtomicBool, u64);
+
+impl<U, T> Future for AllModify<U, T>
+where
+    U: AsUuid,
+{
+    type Output = AllLockModifyGuard<U, T>;
+
+    fn poll(self: Pin<&mut Self>, cxt: &mut Context<'_>) -> Poll<Self::Output> {
+        let (mut lock_record, inner) = self.0.acquire_mutex();
+
+        let wait_type = WaitType::AllWrite;
+        let poll = if lock_record.should_wait(&wait_type, self.2) {
+            lock_record.add_waiter(&self.1, wait_type, cxt.waker().clone(), self.2);
+            Poll::Pending
+        } else {
+            lock_record.add_write_all_lock(self.2);
+            Poll::Ready(AllLockModifyGuard(
+                Arc::clone(&self.0.lock_record),
+                inner.get(),
+                true,
+            ))
+        };
+
+        poll
+    }
+}
+
+impl<U, T> Drop for AllModify<U, T>
+where
+    U: AsUuid,
+{
+    fn drop(&mut self) {
+        let mut lock_record = self
+            .0
+            .lock_record
+            .lock()
+            .expect("Mutex only locked internally");
+        lock_record.cancel(self.2);
+    }
+}
+
+pub struct AllLockModifyGuard<U: AsUuid, T>(Arc<Mutex<LockRecord<U>>>, *mut Table<U, T>, bool);
+
+impl<U, T> Into<AllLockWriteGuard<U, T>> for AllLockModifyGuard<U, T>
```

Comment:
There are a few pieces of the lock code that are not currently used, but may be used in the future. Would you like me to take all of them out?

Reply:
It depends on how sure you are that it will be used in the future. The reason I noticed this is that I was trying to understand the purpose of this particular transformation and I couldn't intuit it and then I couldn't figure it out from the source, 'cause the transformation wasn't used.

Reply:
After reviewing the code, it's possible this will never be used. I'll remove it now.